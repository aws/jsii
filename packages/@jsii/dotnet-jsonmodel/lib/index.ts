import * as path from 'path';

export const repository = path.resolve(__dirname, path.join('..', 'bin', 'Release', 'NuGet'));
