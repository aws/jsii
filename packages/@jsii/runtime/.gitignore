*.js
*.d.ts
dist
webpack
!webpack.config.js

node_modules/
.nyc_output/
coverage/

test/_tarballs/
