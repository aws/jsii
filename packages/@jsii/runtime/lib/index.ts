export * from './host';
export * from './in-out';
