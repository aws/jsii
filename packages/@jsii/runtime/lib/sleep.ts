// We need a shared buffer array for the purpose of this exercise.
const array = new Int32Array(new SharedArrayBuffer(4));

/**
 * Sleeps for a set amount of time, synchronously.
 *
 * @param time the amount of time to wait, in milliseconds.
 */
export function sleep(time: number): void {
  // `Atomics.wait` will block for `time` milliseconds if `array[0]` still has
  // value `0` (which it will, since we just initialized it to that). The return
  // value is irrelevant for our business here.
  Atomics.wait(array, 0, 0, time);
}
