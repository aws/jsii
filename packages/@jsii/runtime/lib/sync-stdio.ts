import * as fs from 'fs';

const STDIN_FD = 0;
const STDOUT_FD = 1;
const STDERR_FD = 2;
const INPUT_BUFFER_SIZE = 1024 * 1024; // not related to max line length

export class SyncStdio {
  private readonly inputQueue = new Array<string>();
  private currentLine = '';

  public writeErrorLine(line: string) {
    this.writeBuffer(Buffer.from(`${line}\n`), STDERR_FD);
  }

  public writeLine(line: string) {
    this.writeBuffer(Buffer.from(`${line}\n`), STDOUT_FD);
  }

  public readLine(): string | undefined {
    if (this.inputQueue.length > 0) {
      return this.inputQueue.shift();
    }

    try {
      const buff = Buffer.alloc(INPUT_BUFFER_SIZE);
      const read = fs.readSync(STDIN_FD, buff, 0, buff.length, null);

      if (read === 0) {
        return undefined;
      }

      const str = buff.slice(0, read).toString();

      for (const ch of str) {
        if (ch === '\n') {
          this.inputQueue.push(this.currentLine);
          this.currentLine = '';
        } else {
          this.currentLine += ch;
        }
      }
    } catch (e) {
      // HACK: node opens STDIN with O_NONBLOCK, meaning attempts to synchrounously read from it may
      // result in EAGAIN. In such cases, the call should be retried until it succeeds. This kind of
      // polling may be terribly inefficient, and the "correct" way to address this is to stop
      // relying on synchronous reads from STDIN. This is however a non-trivial endeavor, and the
      // current state of things is very much broken in node >= 13.2, as can be see in
      // https://github.com/aws/aws-cdk/issues/5187
      if (e.code !== 'EAGAIN') {
        throw e;
      }
    }

    const next = this.inputQueue.shift();
    if (next == null) {
      return this.readLine();
    }

    return next;
  }

  private writeBuffer(buffer: Buffer, fd: number) {
    let offset = 0;
    while (offset < buffer.length) {
      try {
        offset += fs.writeSync(fd, buffer, offset);
      } catch (e) {
        if (e.code !== 'EAGAIN') {
          throw e;
        }
      }
    }
  }
}
