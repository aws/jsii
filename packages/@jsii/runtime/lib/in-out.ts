import { SyncStdio } from './sync-stdio';
import { api } from '@jsii/kernel';

export type Output =
  | { hello: string }
  | { ok: api.KernelResponse }
  | { callback: api.Callback }
  | { pending: true }
  | { error: string; stack?: string };

export type Input =
  | ({ api: string } & api.KernelRequest)
  | { complete: api.CompleteRequest };

export class InputOutput {
  public debug = false;

  private readonly stdio = new SyncStdio();

  public write(obj: Output) {
    const output = JSON.stringify(obj);
    this.stdio.writeLine(output);

    if (this.debug) {
      this.stdio.writeErrorLine(`< ${output}`);
    }
  }

  public read(): Input | undefined {
    let reqLine = this.stdio.readLine();
    if (!reqLine) {
      return undefined;
    }

    // skip recorded responses
    if (reqLine.startsWith('< ')) {
      return this.read();
    }

    // stip "> " from recorded requests
    if (reqLine.startsWith('> ')) {
      reqLine = reqLine.substr(2);
    }

    const input = JSON.parse(reqLine);

    if (this.debug) {
      this.stdio.writeErrorLine(`> ${JSON.stringify(input)}`);
    }

    return input;
  }
}
