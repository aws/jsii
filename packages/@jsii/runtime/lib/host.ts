import {
  api,
  Kernel,
  JsiiFault,
  JsiiError,
  RuntimeError,
  JsiiErrorType,
} from '@jsii/kernel';
import { EventEmitter } from 'events';

import { Input, IInputOutput } from './in-out';

export class KernelHost {
  private readonly kernel = new Kernel(this.callbackHandler.bind(this));
  private readonly eventEmitter = new EventEmitter();

  public constructor(
    private readonly inout: IInputOutput,
    private readonly opts: {
      debug?: boolean;
      debugTiming?: boolean;
      noStack?: boolean;
    } = {},
  ) {
    this.kernel.traceEnabled = opts.debug ?? false;
    this.kernel.debugTimingEnabled = opts.debugTiming ?? false;
  }

  public run() {
    const req = this.inout.read();
    if (!req || 'exit' in req) {
      this.eventEmitter.emit('exit', req?.exit ?? 0);
      return; // done
    }

    this.processRequest(req, () => {
      // Schedule the call to run on the next event loop iteration to
      // avoid recursion.
      setImmediate(() => this.run());
    });
  }

  public once(event: 'exit', listener: (code: number) => void) {
    this.eventEmitter.once(event, listener);
  }

  private callbackHandler(callback: api.Callback) {
    // write a "callback" response, which is a special response that tells
    // the client that there's synchonous callback it needs to invoke and
    // bring back the result via a "complete" request.
    this.inout.write({ callback });

    return completeCallback.call(this);

    function completeCallback(this: KernelHost): void {
      const req = this.inout.read();
      if (!req || 'exit' in req) {
        throw new JsiiFault('Interrupted before callback returned');
      }

      // if this is a completion for the current callback, then we can
      // finally stop this nonsense and return the result.
      const completeReq = req as { complete: api.CompleteRequest };
      if (
        'complete' in completeReq &&
        completeReq.complete.cbid === callback.cbid
      ) {
        if (completeReq.complete.err) {
          if (completeReq.complete.name === JsiiErrorType.JSII_FAULT) {
            throw new JsiiFault(completeReq.complete.err);
          }
          throw new RuntimeError(completeReq.complete.err);
        }

        return completeReq.complete.result;
      }

      // otherwise, process the request normally, but continue to wait for
      // our callback to be completed. sync=true to enforce that `completeCallback`
      // will be called synchronously and return value will be chained back so we can
      // return it to the callback handler.
      return this.processRequest(
        req,
        completeCallback.bind(this),
        callback.sync,
      );
    }
  }

  /**
   * Processes the input request `req` and writes the output response to
   * stdout. This method invokes `next` when the request was fully processed.
   * This either happens synchronously or asynchronously depending on the api
   * (e.g. the "end" api will wait for an async promise to be fulfilled before
   * it writes the response)
   *
   * @param req The input request
   * @param next A callback to invoke to continue
   * @param sync If this is 'true', "next" must be called synchronously. This means
   *             that we won't process any async activity (begin/complete). The kernel
   *             doesn't allow any async operations during a sync callback, so this shouldn't
   *             happen, so we assert in this case to find bugs.
   */
  private processRequest(req: Input, next: () => void, sync = false) {
    if ('callback' in req) {
      throw new JsiiFault(
        'Unexpected `callback` result. This request should have been processed by a callback handler',
      );
    }

    if (!('api' in req)) {
      throw new JsiiFault('Malformed request, "api" field is required');
    }

    const apiReq = req;
    const fn = this.findApi(apiReq.api);

    try {
      const ret = fn.call(this.kernel, req);

      // special case for 'begin' and 'complete' which are on an async
      // promise path. in order to allow the kernel to actually fulfill
      // the promise, and continue any async flows (which may potentially
      // start other promises), we respond only within a setImmediate
      // block, which is scheduled in the same micro-tasks queue as
      // promises. see the kernel test 'async overrides: two overrides'
      // for an example for this use case.
      if (apiReq.api === 'begin' || apiReq.api === 'complete') {
        assertAsyncIsAllowed();

        this.debug('processing pending promises before responding');

        setImmediate(() => {
          this.writeOkay(ret);
          next();
        });

        return undefined;
      }

      // if this is an async method, return immediately and
      // call next only when the promise is fulfilled.
      if (this.isPromise(ret)) {
        assertAsyncIsAllowed();

        this.debug('waiting for promise to be fulfilled');

        const promise = ret;
        promise
          .then((val) => {
            this.debug('promise succeeded:', val);
            this.writeOkay(val);
            next();
          })
          .catch((e) => {
            this.debug('promise failed:', e);
            this.writeError(e);
            next();
          });

        return undefined;
      }

      this.writeOkay(ret);
    } catch (e: any) {
      this.writeError(e);
    }

    // indicate this request was processed (synchronously).
    return next();

    function assertAsyncIsAllowed() {
      if (sync) {
        throw new JsiiFault(
          'Cannot handle async operations while waiting for a sync callback to return',
        );
      }
    }
  }

  /**
   * Writes an "ok" result to stdout.
   */
  private writeOkay(result: any) {
    const res = { ok: result };
    this.inout.write(res);
  }

  /**
   * Writes an "error" result to stdout.
   */
  private writeError(error: JsiiError) {
    const res = {
      error: error.message,
      name: error.name,
      stack: this.opts.noStack ? undefined : error.stack,
    };
    this.inout.write(res);
  }

  /**
   * Returns true if the value is a promise.
   */
  private isPromise(v: any): v is Promise<any> {
    return typeof v?.then === 'function';
  }

  /**
   * Given a kernel api name, returns the function to invoke.
   */
  private findApi(apiName: string): (this: Kernel, arg: Input) => any {
    const fn = (this.kernel as any)[apiName];
    if (typeof fn !== 'function') {
      throw new Error(`Invalid kernel api call: ${apiName}`);
    }
    return fn;
  }

  private debug(...args: any[]) {
    if (!this.opts.debug) {
      return;
    }

    console.error(...args);
  }
}
