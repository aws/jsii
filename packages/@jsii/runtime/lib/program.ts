import * as packageInfo from '../package.json';
import { KernelHost } from './host';
import { InputOutput } from './in-out';

const name = packageInfo.name;
const version = packageInfo.version;

const noStack = !!process.env.JSII_NOSTACK;
const debug = !!process.env.JSII_DEBUG;

const inout = new InputOutput();
const host = new KernelHost(inout, { debug, noStack });

// say hello
inout.write({ hello: `${name}@${version}` });
inout.debug = debug; // we don't want "hello" emitted

host.run();
