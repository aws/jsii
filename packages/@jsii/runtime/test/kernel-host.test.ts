import * as child from 'child_process';
import * as fs from 'fs';
import { api } from '@jsii/kernel';
import * as spec from '@jsii/spec';
import * as path from 'path';
import { KernelHost, InputOutput, Input, Output } from '../lib';

test('can load libraries from within a callback', () => {
  const inout = new TestInputOutput([
    { api: 'load', ...loadRequest('@scope/jsii-calc-base') },
    { api: 'load', ...loadRequest('@scope/jsii-calc-lib') },
    {
      api: 'create',
      fqn: 'Object',
      interfaces: ['@scope/jsii-calc-lib.IFriendly'],
      overrides: [{ method: 'hello' }],
    },
    {
      api: 'invoke',
      objref: { [api.TOKEN_REF]: 'Object@10000' },
      method: 'hello',
    },
    { api: 'load', ...loadRequest('jsii-calc') },
    { complete: { cbid: 'jsii::callback::20000', result: 'SUCCESS!' } },
  ]);
  const host = new KernelHost(inout, { noStack: true, debug: false });
  return new Promise<void>((ok) => {
    host.on('exit', () => ok(inout.expectCompleted()));
    host.run();
  });
});

class TestInputOutput extends InputOutput {
  private readonly inputCommands: Input[];

  public constructor(
    inputCommands: Input[],
    private readonly allowErrors = false,
  ) {
    super();
    this.inputCommands = inputCommands.reverse();
  }

  public read(): Input | undefined {
    return this.inputCommands.pop();
  }

  public write(obj: Output): void {
    if (!this.allowErrors) {
      expect(obj).not.toHaveProperty('error');
    }
    if ('ok' in obj && 'assembly' in obj.ok) {
      // Removing the type count as this is subject to change!
      (obj.ok as any).types = '*redacted*';
    }
    expect(obj).toMatchSnapshot();
  }

  /**
   * Validates that all inputs have been consumed, and all expected outputs have been checked.
   */
  public expectCompleted(): void {
    expect(this.inputCommands).toEqual([]);
  }
}

function loadRequest(library: string): api.LoadRequest {
  const assembly = loadAssembly();
  const tarball = path.join(
    __dirname,
    '_tarballs',
    library,
    `${assembly.fingerprint.replace('/', '_')}.tgz`,
  );
  if (!fs.existsSync(tarball)) {
    packageLibrary(tarball);
  }
  return {
    name: assembly.name,
    version: assembly.version,
    tarball,
  };

  function loadAssembly(): spec.Assembly {
    const assemblyFile = path.resolve(
      require.resolve(`${library}/package.json`),
      '..',
      '.jsii',
    );
    return JSON.parse(fs.readFileSync(assemblyFile, { encoding: 'utf-8' }));
  }

  function packageLibrary(target: string): void {
    const targetDir = path.dirname(target);
    fs.mkdirSync(targetDir, { recursive: true });
    const result = child.spawnSync(
      'npm',
      ['pack', path.dirname(require.resolve(`${library}/package.json`))],
      { cwd: targetDir, shell: true, stdio: ['inherit', 'pipe', 'pipe'] },
    );
    if (result.error) {
      throw result.error;
    }
    if (result.status !== 0) {
      console.error(result.stderr.toString('utf-8'));
      throw new Error(
        `Unable to 'npm pack' ${library}: process ${
          result.signal != null
            ? `killed by ${result.signal}`
            : `exited with code ${result.status}`
        }`,
      );
    }
    fs.renameSync(
      path.join(targetDir, result.stdout.toString('utf-8').trim()),
      target,
    );
  }
}
