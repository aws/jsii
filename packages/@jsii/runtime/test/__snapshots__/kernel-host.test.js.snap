// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can load libraries from within a callback 1`] = `
{
  "ok": {
    "assembly": "@scope/jsii-calc-base-of-base",
    "types": "*redacted*",
  },
}
`;

exports[`can load libraries from within a callback 2`] = `
{
  "ok": {
    "assembly": "@scope/jsii-calc-base",
    "types": "*redacted*",
  },
}
`;

exports[`can load libraries from within a callback 3`] = `
{
  "ok": {
    "assembly": "@scope/jsii-calc-lib",
    "types": "*redacted*",
  },
}
`;

exports[`can load libraries from within a callback 4`] = `
{
  "ok": {
    "$jsii.byref": "Object@10000",
    "$jsii.interfaces": [
      "@scope/jsii-calc-lib.IFriendly",
    ],
  },
}
`;

exports[`can load libraries from within a callback 5`] = `
{
  "callback": {
    "cbid": "jsii::callback::20000",
    "cookie": undefined,
    "invoke": {
      "args": [],
      "method": "hello",
      "objref": {
        "$jsii.byref": "Object@10000",
        "$jsii.interfaces": [
          "@scope/jsii-calc-lib.IFriendly",
        ],
      },
    },
    "sync": true,
  },
}
`;

exports[`can load libraries from within a callback 6`] = `
{
  "ok": {
    "assembly": "jsii-calc",
    "types": "*redacted*",
  },
}
`;

exports[`can load libraries from within a callback 7`] = `
{
  "ok": {
    "result": "SUCCESS!",
  },
}
`;
